File: PHP/answers.php
<?php
/*
 * The data source has a intentional problem in the name column. Where some names
 * have a comma as part of it self. If the programmer is not careful when converting
 * the formats thinks can go wrong.
 */
define('ESCAPE', '"');
define('SEPARATOR', ',');


function extractRowFromLine($line = ''){

  $values = preg_split('/\"\,\"/', $line);

  array_walk($values, function(&$value){

    $value = str_replace(['"'], '', $value);

  });

  return $values;
}

/**
 * Source code here is about items 1 and 2 from read-me.pdf file
 *
 * @param string $filepath
 * @return array
 */
function readJsonToRows($filepath){

  $content = file_get_contents($filepath);
  return json_decode($content, true);
}

/**
 * Source code here is about items 3 and 4 from read-me.pdf file
 *
 * @example curl --location --request GET 'https://economia.awesomeapi.com.br/last/USD-BRL'
 * @return float
 */
function usdToBrl(){

  $curl = curl_init();

  curl_setopt_array($curl, array(
    CURLOPT_URL => 'https://economia.awesomeapi.com.br/last/USD-BRL',
    CURLOPT_RETURNTRANSFER => true,
    CURLOPT_ENCODING => '',
    CURLOPT_MAXREDIRS => 10,
    CURLOPT_TIMEOUT => 0,
    CURLOPT_FOLLOWLOCATION => true,
    CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
    CURLOPT_CUSTOMREQUEST => 'GET',
  ));

  $content = curl_exec($curl);

  curl_close($curl);

  $data = json_decode($content, true);

  return round($data['USDBRL']['bid'], 2);
}

/**
 * Source code here is about item 5 from read-me.pdf file
 *
 * @param array $rows
 * @param float $brl
 * @return array
 */
function updateRowsPrice($rows = array(), $brl = 5.5){

  foreach($rows as &$row){

    $price = $row['price'] * $brl;
    $row['price'] = round($price, 2);
  }


  return $rows;
}

/**
 * Source code here is about item 6 from read-me.pdf file
 *
 * @param array $rows
 * @return array
 */
function createRowsNewColumns($rows = array()){

  foreach($rows as &$row){

    $row['created_at'] = date('Y-m-d H:i:s');
    $row['update_at'] = date('Y-m-d H:i:s');
    $row['delete_at'] = null;
  }

  return $rows;
}

/**
 * Source code here is about items 7, 8, 11 from read-me.pdf file
 *
 * @param array $rows
 * @param string $filepath
 * @return void
 */
function rowsToCsv($rows, $filepath){

  $content = '';

  // Header
  foreach(array_keys($rows[0]) as $key){

    $content .= ESCAPE . $key . ESCAPE . SEPARATOR;
  }

  $content .= PHP_EOL;

  // Body
  foreach($rows as $row){

    foreach($row as &$val){

      $val = ESCAPE . $val . ESCAPE;
    }

    $content .= implode(',', $row) . PHP_EOL;
  }

  file_put_contents($filepath, $content);

  unset($rows);
}

/**
 * Source code here is about items 9 from read-me.pdf file
 *
 * @throws Exception
 * @return array
 */
function csvToRows($filepath){

  if(!file_exists($filepath)){

    throw new \Exception(sprintf("File %s was not found", $filepath));
  }

  $lines = file($filepath, FILE_IGNORE_NEW_LINES|FILE_SKIP_EMPTY_LINES);
  $rows = [];

  // Header
  $keys = extractRowFromLine(array_shift($lines));

  foreach($lines as $line){


    $values = extractRowFromLine($line);

    $rows[] = array_combine($keys, $values);
  }


  return $rows;
}

/**
 * Source code here is about items 10 from read-me.pdf file
 *
 * @return array
 */
function filterRowsForDelete($rows = array()){


  foreach($rows as &$row){

    if($row['price'] >= 275){

      $row['delete_at'] = date('Y-m-d H:i:s');
    }
  }

  return $rows;
}


try {


  $rows = readJsonToRows(__DIR__. '/products.json');
  $brl = usdToBrl();

  if(empty($rows)){

    throw new \Exception('Cannot read products.json');
  }

  $rows = updateRowsPrice($rows, $brl);
  $rows = createRowsNewColumns($rows);

  $filepath = __DIR__. '/products.csv';

  rowsToCsv($rows, $filepath);

  $rows = csvToRows($filepath);


  $rows = filterRowsForDelete($rows);

  rowsToCsv($rows, $filepath);

  printf("success: file %s was created\n", $filepath);

  exit(0);
}
catch(\Exception $e){

  printf("fail: Exception\n");
  printf("message: %s\n", $e->getMessage());
  printf("trace: %s\n", $e->getTraceAsString());
  exit(1);
}

File: Python/answers.py
import sys
import json
import requests
import re
from datetime import datetime

"""
The data source has a intentional problem in the name column. 
Where some names have a comma as part of it self. 
If the programmer is not careful when converting the formats thinks can go wrong.
"""

ESCAPE = '"'
SEPARATOR = ','
EOL = "\n"


def read_json_to_rows(filepath):
    """
    Source code here is about items 1 and 2 from read-me.pdf file

    :parameter filepath str
    :return list
    """
    content = ""

    with open(filepath, "r") as i:
        content += i.read()

    return json.loads(content)


def usd_to_brl():
    """
    Source code here is about items 3 and 4 from read-me.pdf file

    :return float
    """
    response = requests.get('https://economia.awesomeapi.com.br/last/USD-BRL')
    data = response.json()
    return round(float(data['USDBRL']['bid']), 2)


def update_rows_price(rows, brl):
    """
    Source code here is about item 5 from read-me.pdf file

    :parameter rows list
    :parameter brl float
    :return list
    """
    for row in rows:
        price = row['price'] * brl
        row['price'] = round(price, 2)

    return rows


def create_rows_new_columns(rows):
    """
    Source code here is about item 6 from read-me.pdf file

    :parameter rows list
    :return list
    """
    now = datetime.today()

    for row in rows:
        row['created_at'] = now.strftime("%Y-%m-%d %H:%M:%S")
        row['update_at'] = now.strftime("%Y-%m-%d %H:%M:%S")
        row['delete_at'] = ""

    return rows


def rows_to_csv(rows, filepath):
    """
    Source code here is about items 7, 8, 11 from read-me.pdf file

    :parameter rows list
    :parameter filepath str
    :return pass
    """
    content = ""

    # Header
    keys = list(rows[0].keys())
    keys = [ESCAPE + val + ESCAPE for val in keys]
    content += ','.join(keys) + EOL

    # Body
    for row in rows:
        row = [ESCAPE + str(val)  + ESCAPE for val in row.values()]
        content += ','.join(row) + EOL

    file = open(filepath, "w")
    file.write(content)
    file.close()
    pass


def csv_to_rows(filepath):
    """
    Source code here is about items 9 from read-me.pdf file

    :parameter filepath str
    :return list
    """
    regex = re.compile('\"\,\"')
    file = open(filepath, "r")
    lines = file.readlines()
    rows = []

    # Header
    line = lines[0]
    keys = regex.split(line.strip())
    keys = [key.strip('"') for key in keys]

    for line in lines[1:]:
        values = regex.split(line.strip())
        values = [value.strip('"') for value in values]
        rows.append(dict(zip(keys, values)))

    return rows


def filter_rows_for_delete(rows):
    """
    Source code here is about items 1 and 2 from read-me.pdf file

    :parameter filepath str
    :return list
    """
    now = datetime.today()

    for row in rows:
        if float(row['price'])  >= 275:
            row['delete_at'] = now.strftime("%Y-%m-%d %H:%M:%S")

    return rows


try:

    rows = read_json_to_rows("./products.json")
    brl = usd_to_brl()

    if not rows:
        raise "Cannot read products.json"

    rows = update_rows_price(rows, brl)
    rows = create_rows_new_columns(rows)

    filepath = "./products.csv"
    rows_to_csv(rows, filepath)

    rows = csv_to_rows(filepath)

    rows = filter_rows_for_delete(rows)

    rows_to_csv(rows, filepath)

    print("success: file",  filepath, "was created\n")

    exit(0)
except Exception as e:
    print("error:", str(e))
    _, _, traceback = sys.exc_info()
    print("trace:", traceback)
    exit(1)
